docs/generators.md

```diff
@@ -0,0 +1,53 @@
+# Generators
+
+**Generators** are classes that are used to render models for a given language.
+
+## Supported languages
+
+- [JavaScript](../src/generators/javascript/JavaScriptGenerator.ts),
+- [TypeScript](../src/generators/typescript/TypeScriptGenerator.ts),
+- [Java](../src/generators/java/JavaGenerator.ts).
+
+## Generator's options
+
+Each generator extends default options for generator. It means that generator can also have additional options.
+
+Options is passed as first argument to generator's constructor. Check the example:
+
+```ts
+const generator = new TypeScriptGenerator({ ...options });
+```
+
+Default options contains:
+
+| Option | Type | Description | Default value |
+|---|---|---|---|---|
+| `indentation` | Object | Options for indentation. | - |
+| `indentation.type` | String | Type of indentation. Its value can be either `SPACES` or `TABS`. | `SPACES` |
+| `indentation.size` | String | Size of indentation. | 2 |
+| `defaultPreset` | Object | Default preset for generator. For more information, read [customisation](./customisation.md) document. | _Implemented by generator_ |
+| `presets` | Array | Array contains **presets**. For more information, read [customisation](./customisation.md) document. | `[]` |
+
+Below is a list with additional options available for a given generator.
+
+### [TypeScript](../src/generators/typescript/TypeScriptGenerator.ts)
+
+| Option | Type | Description | Default value |
+|---|---|---|---|---|
+| `renderTypes` | Boolean | Renders or not signature for types. | `true` |
+| `modelType` | String | It indicates which model type should be rendered for `object` type. Its value can be either `interface` or `class`. | `class` |
+
+## Custom generator
+
+Creating a new generator which will become a part of the officially supported languages in AsyncAPI Model SDK is pretty simple. The minimum set of required actions to create a new generator are:
+
+- source code must be included in [generators](../src/generators) folder.
+- must extend the abstract [`AbstractGenerator`](../src/generators/AbstractGenerator.ts) class,
+- must define [`Preset`](./customisation.md)'s shape for language,
+- must define language options by passing an interface describing additional options to the first generic argument of [`AbstractGenerator`](../src/generators/AbstractGenerator.ts). Interface must also be extended by `CommonGeneratorOptions` interface,
+- must define default options as static class's field, which must be extended by `defaultGeneratorOptions`,
+- default options must include `defaultPreset` property,
+- must implement `render` function,
+- must define **renderer**s classes for available model types in given language. **Renderer** is an instance of the class with common helper functions to render appropriate model type and must be extended by [`AbstractRenderer`](../src/generators/AbstractRenderer.ts) class - [example](../src/generators/typescript/renderers/ClassRenderer.ts).
```

Comment:
Instead of listing the requirements, what about a here is how you do it 1,2,3? 🤔

Reply:
No matter what to do first, as some items can be done sooner or later, so for me the listing is actually better that list :)

Reply:
It might just be me that don't fancy this way of describing how to implement something 😄 Hope you dont mind, just gonna cc some others to get some more opinions 🙂 @derberg @smoya

Reply:
I kinda agree @jonaslagoni. Getting an idea of the requirements is hard by just reading those points. The reality is that some requirements look way more complicated than in reality. For example:

> Must define language options by passing an interface describing additional options to the first generic argument of [`AbstractGenerator`](../src/generators/AbstractGenerator.ts). The interface must also be extended by `CommonGeneratorOptions` interface

I think this should be part of the library api documentation by its own. Perhaps just showing an example of the usage here would be simpler for the user?

Reply:
@smoya Example is mentioned below the list 

> Check the [generator implementation](../src/generators/typescript/TypeScriptGenerator.ts) for `TypeScript` language to see how it should look like.

Link: https://github.com/asyncapi/generator-model-sdk/blob/master/src/generators/typescript/TypeScriptGenerator.ts

Making inline example like:

```ts
export interface TypeScriptOptions extends CommonGeneratorOptions<TypeScriptPreset> {
  ...
}

export class TypeScriptGenerator extends AbstractGenerator<TypeScriptOptions> {...}
```

is unnecessary in my POV - you have linked example for that. I don't like to write documentation like for idiots -> make this in this way, this in another way etc 😄 I mean, creating guidelines for every case etc. Also I agree with you that `some requirements look way more complicated than in reality`, if you extend custom generator by `AbstractGenerator` your IDE automatically throws errors that you must define `render` function or pass interface as generic argument - and this interface must be extended by CommonGeneratorOptions etc. Some people which will want to contribute, probably won't understand TS, but this is a problem not related to guidelines.